doc/neps/nep-0043-extensible-ufuncs.rst

```diff
@@ -0,0 +1,631 @@
+Note that the ufunc itself can hold on weakly to most ``DTypes`` to allow
+correct cleanup of ``UFuncImpl`` if used correctly.  However, when
+the *output* DTypes are not also input DTypes they have to referenced strongly.
+
+In principle it is be possible to define unbound ``UFuncImpl`` as a full
+Python object.  This will *not* be possible initially, since it would require
+careful duplication of many of the ufunc's metadata on the ``UFuncImpl``
+
+
+Promotion and Discovery
+"""""""""""""""""""""""
+
+NumPy ufuncs are multimethods in the sense that they operate on multiple
+DTypes at once.  While the input (and outpyt) dtypes are attached to numpy
+arrays, the ``ndarray`` type itself does not carry the information of which
+function to apply to the data.
+
+For example, given the input::
+
+    arr1 = np.array([1, 2, 3], dtype=np.int64)
+    arr2 = np.array([1, 2, 3], dtype=np.float64)
+    np.add(arr1, arr2)
+
+has to find the correct ``UfuncImpl`` to perform the operation.
+Ideally, there is an exact match defined, e.g. if the above was written
+as ``np.add(arr1, arr1)``, a ``UFuncImpl[Int64, Int64, out=Int64]`` matches
+exactly can be used.
+However, in the above example there is no direct match, requireing a
+promotion step.
+
+**Implementation:**
+
+1. By default any UFunc has a promotion which uses the common DType of all
+   inputs and tries again.  This is well defined for most mathematical
+   functions, but can be disabled or customized if necessary.
+
+2. Users can *register* new Promoters just as they can register new UFuncImpl.
+   These will use abstract DTypes to allow matching a large variation of
+   signatures.
+   The return value of a promotion function shall be a new ``UFuncImpl``
+   and must consistent over multiple calls with the same input (or return
+   ``NotImplemented`` to indicate an invalid promotion).  This allows
+   caching of the result.
+
+The signature of a promotion function consists is defined by::
+
+    promoter(np.ufunc: ufunc, Tuple[DTypeMeta]: DTypes): -> Union[UFuncImpl, NotImplemented]
+
+Note that DTypes may contain the outputs DType, however, normally the
+output DType should *not* affect which ``UFuncImpl`` is chosen.
+
+In most cases, it should not be necessary to add a custom promotion function,
+however, an example which needs this is for example multiplication with a
+unit.  For example ``timedelta64`` can be multiplied with most integers.
+However, we may only have a loop defined for ``timedelta64 * int64``,
+multiplying with ``int32`` will fail.
+To allow this, the following promoter can be registered for
+``[Timedelta64, Integral, None]``::
+
+    def promote(ufunc, DTypes):
+        res = list(DTypes)
+        try:
+            res[1] = np.common_dtype(DTypes[1], Int64)
+        except TypeError:
+            return NotImplemented
+
+        # Could check that res[1] is actually Int64
+        return ufunc.resolve_impl(tuple(res))
```

Comment:
The mock-up is very useful, but would be even better if `resolve_impl` had already been introduced! (perhaps via the larger mock-up I wondered about above).